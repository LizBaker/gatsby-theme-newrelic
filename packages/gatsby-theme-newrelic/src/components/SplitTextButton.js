import React from 'react';
import PropTypes from 'prop-types';
import { css } from '@emotion/react';
import {
  useTreatments,
  SplitContext,
  useTrack,
} from '@splitsoftware/splitio-react';
import { SPLITS, SPLIT_TRACKING_EVENTS } from '../utils/constants';
import Button from './Button';
import ExternalLink from './ExternalLink';
import useThemeTranslation from '../hooks/useThemeTranslation';

const SplitTextButton = ({ style }) => {
  const { t } = useThemeTranslation();
  const { deven_signupbutton_text } = useTreatments([
    SPLITS.SIGNUP_BUTTON_TEXT,
  ]);
  const track = useTrack();
  const clickCallback = () => {
    track(SPLIT_TRACKING_EVENTS.SIGNUP_BUTTON_CLICK);
  };
  const splitText = deven_signupbutton_text?.treatment;

  const { isReady } = React.useContext(SplitContext);

  return isReady ? (
    <Button
      as={ExternalLink}
      href="https://newrelic.com/signup"
      size={style.size || Button.SIZE.EXTRA_SMALL}
      variant={Button.VARIANT.PRIMARY}
      instrumentation={{ component: 'SplitTextButton' }}
      onClick={clickCallback}
      css={css`
        ${style && style.button}
      `}
    >
      <span>
        {t(splitText === 'start_now' ? 'button.startNow' : 'button.signUp')}
      </span>
    </Button>
  ) : (
    <Button
      as={ExternalLink}
      href="https://newrelic.com/signup"
      size={style.size || Button.SIZE.EXTRA_SMALL}
      variant={Button.VARIANT.PRIMARY}
      instrumentation={{ component: 'SplitTextButton' }}
    >
      <span>{t('button.signUp')}</span>
    </Button>
  );
};
SplitTextButton.propTypes = {
  style: PropTypes.object,
};

export default SplitTextButton;
