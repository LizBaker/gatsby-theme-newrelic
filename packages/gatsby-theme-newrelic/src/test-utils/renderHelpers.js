import React from 'react';
import { render } from '@testing-library/react';
import { themeNamespace } from '../utils/defaultOptions';
import { I18nextProvider } from 'react-i18next';
import {
  LocationProvider,
  createHistory,
  createMemorySource,
} from '@reach/router';
import LocaleProvider from '../components/LocaleProvider';
import translations from '../i18n/translations/en.json';
import i18n from 'i18next';

export const renderWithTranslation = (component, options) => {
  i18n.init({
    defaultNS: 'translation',
    initImmediate: false,
    fallbackLng: 'en',
    lng: 'en',
    ns: [themeNamespace, 'translation'],
    resources: {
      en: {
        [themeNamespace]: translations,
      },
    },
    interpolation: {
      escapeValue: false,
    },
    react: {
      useSuspense: false,
    },
  });

  return render(
    <I18nextProvider i18n={i18n}>
      <LocaleProvider i18n={i18n}>{component}</LocaleProvider>
    </I18nextProvider>,
    options
  );
};

export const renderWithProviders = (
  component,
  { locale = 'en', ...options } = {}
) => {
  i18n.init({
    defaultNS: 'translation',
    initImmediate: false,
    fallbackLng: 'en',
    lng: locale,
    ns: [themeNamespace, 'translation'],
    resources: {
      en: {
        [themeNamespace]: translations,
      },
    },
    interpolation: {
      escapeValue: false,
    },
    react: {
      useSuspense: false,
    },
  });

  const history = createHistory(createMemorySource('/'));

  return render(
    <LocationProvider history={history}>
      <I18nextProvider i18n={i18n}>
        <LocaleProvider i18n={i18n}>{component}</LocaleProvider>
      </I18nextProvider>
    </LocationProvider>,
    options
  );
};
