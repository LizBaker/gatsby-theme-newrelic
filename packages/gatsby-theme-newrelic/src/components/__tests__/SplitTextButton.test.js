import React from 'react';
import TestRenderer from 'react-test-renderer';
import rerender from '@testing-library/react';
import { SplitFactory } from '@splitsoftware/splitio-react';
import SplitTextButton from '../SplitTextButton';
import { renderWithProviders } from '../../test-utils/renderHelpers';
import { screen } from '@testing-library/react';

import useThemeTranslation from '../../hooks/useThemeTranslation';
import Button from '../Button';

// jest.mock('../../hooks/useThemeTranslation', () =>
//   jest.fn(() => ({ t: jest.fn((s) => s) }))
// );

// jest.mock('../Button', () => ({ children }) => <button>{children}</button>);

// jest.mock('gatsby', () => ({
//   __esModule: true,
//   graphql: () => {},
//   Link: ({ to, ...props }) => <a href={to} {...props} />,
//   useStaticQuery: () => ({
//     allLocale: {
//       nodes: [
//         {
//           name: 'English',
//           locale: 'en',
//           localizedPath: '/en',
//           isDefault: true,
//         },
//       ],
//     },
//     site: {
//       siteMetadata: {
//         siteUrl: 'https://github.com/foo/bar',
//         repository: 'https://foobar.net',
//       },
//     },
//     newRelicThemeConfig: {
//       tessen: {
//         product: 'foo',
//         subproduct: 'foobar',
//       },
//     },
//   }),
// }));

export const renderWithTranslation = (component, options) => {
  i18n.init({
    ...i18nextOptions,
    lng: 'en',
    resources: {
      en: {
        [themeNamespace]: translations,
      },
    },
  });

  return render(
    <I18nextProvider i18n={i18n}>
      <LocaleProvider i18n={i18n}>{component}</LocaleProvider>
    </I18nextProvider>,
    options
  );
};

test('renders default button text with no split', async () => {
  const config = createSplitConfig('start_now');
  //   console.log(config);

  const element = renderWithProviders(
    <SplitFactory config={config} updateOnSdkTimedout={true}>
      <SplitTextButton />
    </SplitFactory>
  );

  await TestRenderer.act(async () => {
    rerender(element);
  });

  expect(screen.getByText('Free account')).toBeTruthy();
});

const createSplitConfig = (text) => {
  return {
    core: {
      authorizationKey: 'localhost',
      key: 'user',
    },
    features: {
      deven_signupbutton_text:
        text === 'start_now'
          ? { treatment: 'start_now', config: '{ "text": "Start now" }' }
          : { treatment: 'free_account', config: '{ "text": "Free account" }' },
    },
    scheduler: {
      offlineRefreshRate: 15,
    },
    debug: false,
  };
};
