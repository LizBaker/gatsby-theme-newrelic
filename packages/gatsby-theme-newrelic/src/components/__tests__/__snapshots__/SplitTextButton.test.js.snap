// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders default button with no split 1`] = `null`;
